File: CSharpCodeBase/components/thing/dropmanagercomponent.lua
local Class = require 'utils.hump.class'
local Component = require 'components.component'
local ConfigComponent = require 'components.services.config'

local DropManagerComponent = Class({__includes=Component, init=function(self, entity)
      Component.init(self, entity)
      self.config = ConfigComponent('drop', self.entity.name)
      self.entity.dropManager = self
      
      self.healthCheckPoints = {}
      self.healthStep = {}
      self:LoadDrop(self.entity.config:Get('defaultDrop') or 'none')
end})

function DropManagerComponent:LoadDrop(entityName)
  if entityName == 'none' or entityName == nil then
    self.isActive = false
    return 
  else
    self.isActive = true
  end
  self.dropItems = self.config:Get("drop_item_" .. entityName)
  self.dropEvents = self.config:Get("drop_event_" .. entityName)
  self.dropPerlifes = self.config:Get("drop_perlife_" .. entityName)
  self.dropChance = self.config:Get("drop_chance_" .. entityName)
  self.healthCheckPoints = {}
  self.healthStep = {}
  
  for i=1, #self.dropEvents, 1 do
    table.insert(self.healthStep,(self.entity.health.maxAmount/100)*self.dropPerlifes[i])
    table.insert(self.healthCheckPoints, self.entity.health.maxAmount - (self.entity.health.maxAmount/100)*self.dropPerlifes[i])
  end
end

function DropManagerComponent:DropAfterHit()
  if self.isActive == false then return end
  for i=1, #self.dropEvents, 1 do
    if self.dropEvents[i] == "hit" then
      local count = 0;
      count = self:Drop(i, count)
      if count ~=0 then
        -- вставить выпадание дропа
      end
    end
  end
end

function DropManagerComponent:DropAfterDeath()
  if self.isActive == false then return end
  self:DropAfterHit()
  for i=1, #self.dropEvents, 1 do
    if self.dropEvents[i] == "death" then
      if self:IsDrop(i) then
        -- вставить выпадание дропа
      end
    end
  end
end

function DropManagerComponent:Drop(i, dropCount)
  if self.healthCheckPoints[i] >= self.entity.health.amount then
    self.healthCheckPoints[i] = self.healthCheckPoints[i] - self.healthStep[i]
    if self:IsDrop(i) then
      dropCount = dropCount + 1
    end
    dropCount = self:Drop(i, dropCount)
  end
  return dropCount
end

function DropManagerComponent:IsDrop(i)
  if math.chance(self.dropChance[i]) then
    return true
  end
  return false
end

function DropManagerComponent:OnRegeneration()
  if self.isActive == false then return end
  for i=1, #self.healthCheckPoints, 1 do
    if self.healthCheckPoints[i] <= self.entity.health.amount then
      self.healthCheckPoints[i] = self.healthCheckPoints[i] + self.healthStep[i]
    end
  end
end

return DropManagerComponent

File: CSharpCodeBase/components/thing/dropmanagercomponent.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class dropmanagercomponent :component{
 public void init(self, entity);
       Component.init(self, entity);
       this.config = ConfigComponent("drop", this.entity.name);
       this.entity.dropManager = self;
       
       this.healthCheckPoints = {}
       this.healthStep = {}
       self:LoadDrop(this.entity.config:Get("defaultDrop")  ||  "none");
 }})
 public void LoadDrop(entityName){
   if(entityName == "none"  ||  entityName == null  ){ 
     this.isActive = false;
     return ;
   }else{
     this.isActive = true;
   }
   this.dropItems = this.config:Get("drop_item_" .. entityName);
   this.dropEvents = this.config:Get("drop_event_" .. entityName);
   this.dropPerlifes = this.config:Get("drop_perlife_" .. entityName);
   this.dropChance = this.config:Get("drop_chance_" .. entityName);
   this.healthCheckPoints = {}
   this.healthStep = {}
   
   for(i=1, #this.dropEvents, 1 ){
     table.insert(this.healthStep,(this.entity.health.maxAmount/100)*this.dropPerlifes[i]);
     table.insert(this.healthCheckPoints, this.entity.health.maxAmount - (this.entity.health.maxAmount/100)*this.dropPerlifes[i]);
   }
 }
 public void DropAfterHit(){
   if(this.isActive == false  ){  return }
   for(i=1, #this.dropEvents, 1 ){
     if(this.dropEvents[i] == "hit"  ){ 
       var count = 0;
       count = self:Drop(i, count);
       if(count !=0  ){ 
         // вставить выпадание дропа
       }
     }
   }
 }
 public void DropAfterDeath(){
   if(this.isActive == false  ){  return }
   self:DropAfterHit();
   for(i=1, #this.dropEvents, 1 ){
     if(this.dropEvents[i] == "death"  ){ 
       if(self:IsDrop(i)  ){ 
         // вставить выпадание дропа
       }
     }
   }
 }
 public void Drop(i, dropCount){
   if(this.healthCheckPoints[i] >= this.entity.health.amount  ){ 
     this.healthCheckPoints[i] = this.healthCheckPoints[i] - this.healthStep[i];
     if(self:IsDrop(i)  ){ 
       dropCount = dropCount + 1;
     }
     dropCount = self:Drop(i, dropCount);
   }
   return dropCount;
 }
 public void IsDrop(i){
   if(math.chance(this.dropChance[i])  ){ 
     return true;
   }
   return false;
 }
 public void OnRegeneration(){
   if(this.isActive == false  ){  return }
   for(i=1, #this.healthCheckPoints, 1 ){
     if(this.healthCheckPoints[i] <= this.entity.health.amount  ){ 
       this.healthCheckPoints[i] = this.healthCheckPoints[i] + this.healthStep[i];
     }
   }
 }
 }}